File: frontend/src/api/sprints.ts
import { BACKEND_URL, requestJson, throwIfErrorResponse } from '../../lib/api';

export interface Sprint {
  id: string;
  name: string;
  goal: string;
  start_date: string;
  end_date: string;
  status: string;
  created_at: string;
}

export interface Column {
  id: string;
  sprint_id: string;
  name: string;
  order_index: number;
}

export interface TaskMessage {
  id: string;
  task_id: string;
  sender_id: string;
  sender_name: string;
  sender_type: string;
  message_type: string;
  content: string;
  created_at: string;
}

export interface SprintRunStep {
  id: string;
  run_id: string;
  title: string;
  status: 'todo' | 'in_progress' | 'done' | 'blocked' | 'cancelled' | string;
  order_index: number;
  created_at: string;
  updated_at: string;
}

export interface SprintRun {
  id: string;
  task_id: string;
  employee_id: string;
  employee_name: string;
  status: 'running' | 'done' | 'failed' | string;
  summary: string;
  error: string;
  created_at: string;
  updated_at: string;
  steps: SprintRunStep[];
}

export interface Task {
  id: string;
  sprint_id: string;
  column_id: string;
  title: string;
  requirements: string;
  description: string;
  priority: string;
  employee_id: string;
  created_at: string;
  updated_at: string;
  messages: TaskMessage[];
  runs: SprintRun[];
}

export interface SprintData {
  sprint: Sprint;
  columns: Column[];
  tasks: Task[];
}

export async function getActiveSprint(): Promise<SprintData> {
  const res = await fetch(`${BACKEND_URL}/api/sprints/active`, {
    headers: {
      Authorization: `Bearer ${localStorage.getItem('auth_token') || ''}`,
    },
  });
  if (!res.ok && res.status === 404) {
    throw new Error('No active sprint found');
  }
  await throwIfErrorResponse(res, 'The active sprint could not be loaded.');
  return res.json();
}

export async function createTask(sprintId: string, data: { title: string; column_id: string; requirements?: string; description?: string; priority?: string; employee_id?: string }): Promise<{ task_id: string }> {
  return requestJson<{ task_id: string }>(`/api/sprints/${sprintId}/tasks`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('auth_token') || ''}`,
    },
    body: JSON.stringify(data),
  }, 'The sprint task could not be created.');
}

export async function updateTaskColumn(taskId: string, columnId: string): Promise<void> {
  await requestJson<{ status: string }>(`/api/sprints/tasks/${taskId}/column`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('auth_token') || ''}`,
    },
    body: JSON.stringify({ column_id: columnId }),
  }, 'The task column could not be updated.');
}

export async function updateTaskDetails(taskId: string, data: { title: string; description: string; requirements: string }): Promise<void> {
  await requestJson<{ status: string }>(`/api/sprints/tasks/${taskId}/details`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('auth_token') || ''}`,
    },
    body: JSON.stringify(data),
  }, 'The task details could not be updated.');
}

export async function updateTaskAssignment(taskId: string, employeeId: string): Promise<void> {
  await requestJson<{ status: string }>(`/api/sprints/tasks/${taskId}/assignment`, {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('auth_token') || ''}`,
    },
    body: JSON.stringify({ employee_id: employeeId }),
  }, 'The task assignment could not be updated.');
}

export async function runTask(sprintId: string, taskId: string): Promise<{ run_id: string; status: string }> {
  return requestJson<{ run_id: string; status: string }>(`/api/sprints/${sprintId}/tasks/${taskId}/run`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('auth_token') || ''}`,
    },
  }, 'The sprint task run could not be started.');
}

export async function addTaskMessage(taskId: string, data: { content: string; sender_id?: string; sender_name?: string; sender_type?: string; message_type?: string }): Promise<{ message_id: string }> {
  return requestJson<{ message_id: string }>(`/api/sprints/tasks/${taskId}/messages`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${localStorage.getItem('auth_token') || ''}`,
    },
    body: JSON.stringify(data),
  }, 'The task message could not be added.');
}

File: backend/api/routers/sprints.py
from fastapi import APIRouter, HTTPException, Depends
from typing import List, Dict, Any, Optional
from pydantic import BaseModel

from backend.core.errors import AppError
from backend.db.stores.sprint_store import SprintStore
from backend.lib.auth.session import get_current_user
from backend.core.dispatch.jobs import enqueue_job, QueueFullError

router = APIRouter(prefix="/sprints")

def get_store():
    return SprintStore()

class CreateTaskRequest(BaseModel):
    title: str
    column_id: str
    requirements: str = ""
    description: str = ""
    priority: str = "medium"
    employee_id: str = ""

class UpdateTaskColumnRequest(BaseModel):
    column_id: str

class UpdateTaskDetailsRequest(BaseModel):
    title: str
    description: str
    requirements: str
    priority: Optional[str] = None

class UpdateTaskAssignmentRequest(BaseModel):
    employee_id: str

class AddMessageRequest(BaseModel):
    content: str
    sender_id: str = ""
    sender_name: str = ""
    sender_type: str = "human"
    message_type: str = "user"


@router.get("/active")
async def get_active_sprint(store: SprintStore = Depends(get_store), user=Depends(get_current_user)):
    owner_id = (user or {}).get("id", "") if isinstance(user, dict) else ""
    sprint = store.get_active_sprint(owner_id)
    if not sprint:
        raise HTTPException(status_code=404, detail="No active sprint found")

    columns = store.get_sprint_columns(sprint["id"])
    tasks = store.get_sprint_tasks(sprint["id"])

    for task in tasks:
        task["messages"] = store.get_task_messages(task["id"])
        task["runs"] = store.list_runs_for_task(task["id"])

    return {
        "sprint": sprint,
        "columns": columns,
        "tasks": tasks
    }

@router.post("/{sprint_id}/tasks")
async def create_task(sprint_id: str, req: CreateTaskRequest, store: SprintStore = Depends(get_store), user=Depends(get_current_user)):
    owner_id = (user or {}).get("id", "") if isinstance(user, dict) else ""
    sprint = store.get_sprint(sprint_id)
    if not sprint or (owner_id and sprint.get("owner_id", "") not in (owner_id, "")):
        raise HTTPException(status_code=404, detail="Sprint not found")
    task_id = store.create_task(
        sprint_id=sprint_id,
        column_id=req.column_id,
        title=req.title,
        requirements=req.requirements,
        description=req.description,
        priority=req.priority,
        employee_id=req.employee_id
    )
    return {"task_id": task_id}

@router.patch("/tasks/{task_id}/column")
async def update_task_column(task_id: str, req: UpdateTaskColumnRequest, store: SprintStore = Depends(get_store), user=Depends(get_current_user)):
    store.update_task_column(task_id, req.column_id)
    return {"status": "success"}

@router.patch("/tasks/{task_id}/details")
async def update_task_details(task_id: str, req: UpdateTaskDetailsRequest, store: SprintStore = Depends(get_store), user=Depends(get_current_user)):
    store.update_task_details(task_id, req.title, req.description, req.requirements, req.priority)
    return {"status": "success"}

@router.patch("/tasks/{task_id}/assignment")
async def update_task_assignment(task_id: str, req: UpdateTaskAssignmentRequest, store: SprintStore = Depends(get_store), user=Depends(get_current_user)):
    store.update_task_assignment(task_id, req.employee_id)
    return {"status": "success"}

@router.post("/{sprint_id}/tasks/{task_id}/run")
async def run_task(
    sprint_id: str,
    task_id: str,
    store: SprintStore = Depends(get_store),
    user=Depends(get_current_user),
):
    sprint = store.get_sprint(sprint_id)
    if not sprint:
        raise HTTPException(status_code=404, detail=f"Sprint {sprint_id} not found")

    tasks = store.get_sprint_tasks(sprint_id)
    task = next((t for t in tasks if t["id"] == task_id), None)
    if not task:
        raise HTTPException(status_code=404, detail=f"Task {task_id} not found")
    employee_id = (task.get("employee_id") or "").strip()
    if not employee_id:
        raise HTTPException(status_code=400, detail="Task has no assigned employee")

    user_id = (user or {}).get("id") if isinstance(user, dict) else "anonymous"
    owner_id = user_id or ""

    try:
        job = await enqueue_job(
            kind="sprint",
            lane="sprint",
            user_id=user_id or "anonymous",
            owner_id=owner_id,
            required_employee_ids=[employee_id],
            payload={"sprint_id": sprint_id, "task_id": task_id},
        )
    except QueueFullError as e:
        raise AppError(
            code="SPRINT_TASK_RUN_QUEUE_FULL",
            message=str(e),
            status_code=429,
            retryable=True,
            details={"task_id": task_id, "sprint_id": sprint_id},
        ) from e

    return {"job_id": job.id, "status": "queued", "employee_id": employee_id}


@router.post("/tasks/{task_id}/messages")
def add_task_message(task_id: str, req: AddMessageRequest, store: SprintStore = Depends(get_store)):
    msg_id = store.add_task_message(
        task_id=task_id,
        content=req.content,
        sender_id=req.sender_id,
        sender_name=req.sender_name,
        sender_type=req.sender_type,
        message_type=req.message_type
    )
    return {"message_id": msg_id}
